### PR TITLE
add support for the asymmetric Encrypt and Decrypt restful API

### PR DESCRIPTION
**Pull request by Liang1XZhao**
this restful API will exposed to public as:
- POST <ehsm_srv_address>/ehsm?Action=AsymmetricEncrypt
- POST <ehsm_srv_address>/ehsm?Action=AsymmetricDecrypt
add the corresponding test cases.

change server port 3000 to 9000.
Napi_* Method from ehsm_kms_server.JS extract to ehsm_napi.js

test: passed the unittest

Signed-off-by: Liang1XZhao <liang1x.zhao@intel.com>

### CODE REVIEW

File: core/App/ehsm_napi.h [hunk 1]
```diff
@@ -92,6 +92,26 @@ typedef struct {
         retString += "}";
         return StringToChar(retString);
 	};
+
+    static int getCode(char* jsonChar){
+        std::string jsonString = jsonChar;
```
Comment (yang8621) [inline core/App/ehsm_napi.h:97]: if NULL, will be crashed or not ?

Comment (Liang1XZhao) [reply core/App/ehsm_napi.h:97]: will be crashed

File: core/App/ehsm_core_test.cpp [hunk 2]
```diff
@@ -126,144 +127,52 @@ void test_AES128()
...
+    returnJsonChar = NAPI_AsymmetricDecrypt(cmk_base64, ciphertext_base64);
+    if(RetJsonObj::getCode(returnJsonChar) != 200){
+        printf("NAPI_AsymmetricDecrypt failed, error message: %s \n", RetJsonObj::getMessage(returnJsonChar).c_str());
         goto cleanup;
     }
-    printf("Verify signature SUCCESSFULLY!\n");
+    printf("NAPI_AsymmetricDecrypt json : %s\n", returnJsonChar);
+    plaintext_base64 = RetJsonObj::readData_string(returnJsonChar, "plaintext_base64");
+    printf("Decrypted plaintext : %s\n",base64_decode(plaintext_base64).c_str());
+    printf("NAPI_AsymmetricDecrypt data SUCCESSFULLY!\n");
```
Comment (yang8621) [inline core/App/ehsm_core_test.cpp:167]: Need to compare the original plaintext and the decrypted text.

File: core/App/ehsm_napi.h (hunk 1 above)
Comment (yang8621) [reply core/App/ehsm_napi.h:97]: So please return an error code if NULL.

File: core/App/ehsm_napi.h [hunk 3]
```diff
@@ -92,6 +92,26 @@ typedef struct {
...
+        int startIndex = jsonString.find("\"code\"") + strlen("\"code\":");
+        int endIndex = jsonString.find_first_of(",", startIndex);
+
+        std::string code_str = jsonString.substr(startIndex, (endIndex - startIndex));
+        
+        return atoi(code_str.c_str());
+    }
+
+    static std::string getMessage(char* jsonChar){
+        std::string jsonString = jsonChar;
```
Comment (yang8621) [inline core/App/ehsm_napi.h:108]: same here.

File: ehsm_kms_service/test/test_kms_with_rest.py [hunk 4]
```diff
@@ -28,66 +28,114 @@ def test_params(payload):
...
+            "cmk_base64": json.loads(create_res.text)['result']['cmk_base64'],
+            "ciphertext_base64": json.loads(asymmetricEncrypt_res.text)['result']['ciphertext_base64']
+        })
+    print('AsymmetricDecrypt req:\n%s\n' %(params))
+    asymmetricDecrypt_res = requests.post(url=base_url + "AsymmetricDecrypt", data=json.dumps(params), headers=headers)
+    print('AsymmetricDecrypt resp:\n%s\n' %(asymmetricDecrypt_res.text))
+    if(check_result(asymmetricDecrypt_res, 'AsymmetricDecrypt', 'test_RSA3072_encrypt_decrypt') == False):
+        return
+    plaintext = str(base64.b64decode(json.loads(asymmetricDecrypt_res.text)['result']['plaintext_base64']), 'utf-8').strip(b"\x00".decode())
+    print('AsymmetricDecrypt plaintext:\n%s\n' %(plaintext))
```
Comment (yang8621) [inline ehsm_kms_service/test/test_kms_with_rest.py:64]: still need to compare the string.

File: core/App/ehsm_napi.h (hunk 1 above)
Comment (Liang1XZhao) [reply core/App/ehsm_napi.h:97]: Will be fix in next PR.

File: ehsm_kms_service/ehsm_kms_server.js [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
 const crypto = require('crypto')
 const app = express();
 const logger = require('./logs/logger');
+const ehsm_napi = require('./ehsm_napi');
 const { ehsm_kms_params, ehsm_keyspec_t, ehsm_keyorigin_t} = require('./ehsm_kms_params.js')
 
 app.use(express.json());
 
-const port = 3000;
+const port = 9000;
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_server.js:10]: didn't see you change the dockerfile， how can it pass the test? due to the docker still expose the 3000.

and also please change the build script, and change the "-P" to make it map to the dedicate port

File: ehsm_kms_service/ehsm_kms_params.js [hunk 6]
```diff
@@ -106,7 +106,28 @@ const ehsm_kms_params = {
       maxLength: _6kb_length,
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_params.js:106]: so do you want to change the above length check? they are also not correct as i told you previously.

Comment (syan10) [reply ehsm_kms_service/ehsm_kms_params.js:106]: or you can do them in another patch for the code refine

File: ehsm_kms_service/test/test_kms_with_rest.py [hunk 7]
```diff
@@ -28,66 +28,114 @@ def test_params(payload):
...
+def test_Stest_RSA3072_sign_verify(base_url, headers):
+    print('====================test_Stest_RSA3072_sign_verify start===========================')
+    params=test_params({
+            "keyspec": "EH_RSA_3072",
+            "origin": "EH_INTERNAL_KEY"
+        })
+    print('CreateKey req:\n%s\n' %(params))
+    create_res = requests.post(url=base_url + "CreateKey", data=json.dumps(params), headers=headers)
+    print('CreateKey resp:\n%s\n' %(create_res.text))
+    if(check_result(create_res, 'CreateKey', 'test_Stest_RSA3072_sign_verify') == False):
```
Comment (syan10) [inline ehsm_kms_service/test/test_kms_with_rest.py:76]: why not put the check_result() before print out the response?

exchange the line#75 and line#76, and same with below...

File: ehsm_kms_service/ehsm_kms_server.js (hunk 5 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/ehsm_kms_server.js:10]: It will be fix

File: ehsm_kms_service/ehsm_kms_params.js (hunk 6 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/ehsm_kms_params.js:106]: It will be fix in another patch.

File: ehsm_kms_service/test/test_kms_with_rest.py (hunk 7 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/test/test_kms_with_rest.py:76]: It will be fix

**Review by syan10 — COMMENTED** [review]
LGTM

**Pull request closed without merging**